### PR TITLE
Spectral fractional PDE in MFEM

### PR DESCRIPTION
**Pull request by brendankeith**
This PR is the first step towards providing support for PDE-based sampling of Matern random fields [4]. Such fields will be used in LDRD 22-ERD-009 to simulate uncertainties present in the additive manufacturing process [2,3].

In this PR, we provide an MFEM example that solves the following fractional PDE:

      ( - Δ )^α u = f  in Ω,      u = 0  on ∂Ω,      α > 0,

To solve this FPDE, we rely on a rational approximation of the normal linear operator `A^{-α}`, where `A = - Δ` (with associated homogenous boundary conditions). Namely, we first *approximate the operator*

      A^{-α} ≈ Σ_{i=0}^N c_i (A + d_i I)^{-1},      d_0 = 0,   d_i > 0,    i=1,...,N,

where `I` is the L2-identity operator and the coefficients `c_i` and `d_i` are generated offline to a prescribed accuracy in a pre-processing step. We use the triple-A algorithm [0] to generate the rational approximation that this partial fraction expansion derives from. We then solve N+1 independent integer-order PDEs,

      A u_i + d_i u_i = c_i f  in Ω,      u_i = 0  on ∂Ω,      i=0,...,N,

using the finite element method and sum `u_i` to arrive at an approximate solution of the FPDE

      u ≈ Σ_{i=0}^N u_i.

### NOTES:

* This technique is quite general and will provide discrete solutions to FPDEs other than the fractional Laplacian.
* Fractional anisotropic diffusion will be used in the upcoming PR on PDE-based sampling of Matern random fields.

### References:

[0] Nakatsukasa, Y., Sète, O., & Trefethen, L. N. (2018). The AAA algorithm for rational approximation. SIAM Journal on Scientific Computing, 40(3), A1494-A1522.

[1] Harizanov, S., Lazarov, R., Margenov, S., Marinov, P., & Pasciak, J. (2020). Analysis of numerical methods for spectral fractional elliptic equations based on the best uniform rational approximation. Journal of Computational Physics, 408, 109285.

[2] Khristenko, U., Constantinescu, A., Tallec, P. L., Oden, J. T., & Wohlmuth, B. (2020). A statistical framework for generating microstructures of two-phase random materials: application to fatigue analysis. Multiscale Modeling & Simulation, 18(1), 21-43.

[3] Khristenko, U., Constantinescu, A., Tallec, P. L., & Wohlmuth, B. (2021). Statistically equivalent surrogate material models and the impact of random imperfections on elasto-plastic response. arXiv preprint arXiv:2112.06655.

[4] Lindgren, F., Bolin, D., & Rue, H. (2022). The SPDE approach for Gaussian and non-Gaussian fields: 10 years and still running. Spatial Statistics, 100599.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2832](https://github.com/mfem/mfem/pull/2832) | @brendankeith | @tzanio | @bslazarov + @jamiebramwell + @YohannDudouit | 3/8/22 | 4/12/22 | 4/18/22 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
A few relevant items from the [PR checklist](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-request-checklist) copied here for convenience:

- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.

**Comment by brendankeith** [comment]
@bslazarov, @jamiebramwell Can you take a look at this PR before our group meeting next week?

**Comment by brendankeith** [comment]
@bslazarov @jamiebramwell just a reminder about reviewing this PR!

File: examples/ex33.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      u+=x;
+
+      // 16. Send the solutions by socket to a GLVis server.
+      if (visualization)
+      {
+         xout << "solution\n" << mesh << x << flush;
+         uout << "solution\n" << mesh << u << flush;
+      }
+   }
+   return 0;
```
Comment (bslazarov) [inline examples/ex33.cpp:178]: delete fec;  - the finite element collection is not deleted

File: examples/ex33.hpp [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+//       Computing, 40(3), A1494-A1522.
+
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
```
Comment (bslazarov) [inline examples/ex33.hpp:42]: All functions should be documented.

File: examples/ex33.hpp [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      }
+
+      Vector verr(val);
+      verr-=R;
+
+      if (verr.Normlinf() <= tol*val.Normlinf()) { break; }
+   }
+}
+
+void ComputePolesAndZeros(const Vector &z, const Vector &f, const Vector &w,
```
Comment (bslazarov) [inline examples/ex33.hpp:155]: Documentation

File: examples/ex33.hpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+      for (int k=0; k<psize; k++)
+      {
+         if (k != i) { tmp_denom *= poles[i]-poles[k]; }
+      }
+      coeffs[i] *= tmp_numer / tmp_denom;
+   }
+}
+
+
+// x^-a
```
Comment (bslazarov) [inline examples/ex33.hpp:244]: Documentation?

File: examples/ex33.hpp [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      }
+   }
+#else
+   mfem_error("Compiled without LAPACK");
+#endif
+
+   scale = w * f / w.Sum();
+}
+
+
```
Comment (bslazarov) [inline examples/ex33.hpp:217]: Add documentation.

File: examples/ex33.hpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+      coeffs = Array<double>({0.0292386, 0.0143338, 0.0109015, 0.00976958,
+                              0.00943417, 0.00948077, 0.00985799, 0.010889,
+                              0.0138588, 0.0263289, 0.96953});
+      poles = Array<double> ({-10085.1, -1652.6, -524.342, -199.521,
+                              -80.2328, -32.3926, -12.7148, -4.63855,
+                              -1.39884, -0.221076, 0.});
+   }
+   else
+   {
+      if (alpha != 0.5)
```
Comment (bslazarov) [inline examples/ex33.hpp:276]: Why these values are used by default? Aren't they valid for 0.5? If so just write in the documentation that the code is implemented for 0.33, 0.99 and 0.5.

File: examples/ex33.hpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+                                         double lmin = 1., double lmax = 1000.,
+                                         double tol=1e-10,  int npoints = 1000,
+                                         int max_order = 100)
+{
+#ifndef MFEM_USE_LAPACK
+   mfem::out
+         << "MFEM is compiled without LAPACK. Using precomputed PartialFractionApproximation"
+         << std::endl;
+
+   if (alpha == 0.33)
```
Comment (bslazarov) [inline examples/ex33.hpp:256]: Float comparison - use eps.

File: examples/ex33.hpp [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+   {
+      coeffs = Array<double> ({2002.55,  99.7691,  29.0575,  12.9842,
+                               6.90263,  3.96688,  2.36013,  1.42565,
+                               0.867623, 0.529436, 0.317975, 0.0891797});
+      poles = Array<double> ({-47928.,  -3451.45, -996.636, -388.58,
+                              -168.228, -75.3162, -33.6235, -14.5403,
+                              -5.84563, -1.9975, -0.434608,  0.});
+   }
+   else if (alpha == 0.99)
```
Comment (bslazarov) [inline examples/ex33.hpp:265]: Float comparison

File: examples/ex33.hpp (hunk 6 above)
Comment (bslazarov) [inline examples/ex33.hpp:276]: Float comparison.

File: examples/ex33.hpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+      {
+         if (k != i) { tmp_denom *= poles[i]-poles[k]; }
+      }
+      coeffs[i] *= tmp_numer / tmp_denom;
+   }
+}
+
+
+// x^-a
+void ComputePartialFractionApproximation(double alpha,
```
Comment (bslazarov) [inline examples/ex33.hpp:245]: Some guidance on how to select the default parameters is necessary. A reference should be added to the paper describing the approximation.

**Review by bslazarov — COMMENTED** [review]
Tested with cmake and valgrind - works fine apart from a small memory leak. Documentation for the implemented functions is missing.

File: examples/ex33.hpp (hunk 2 above)
Comment (brendankeith) [reply examples/ex33.hpp:42]: Done.

File: examples/ex33.hpp (hunk 6 above)
Comment (brendankeith) [reply examples/ex33.hpp:341]: Done

**Comment by brendankeith** [comment]
Thanks, @bslazarov. I just went through your comments.

**Review by bslazarov — APPROVED** [review]

File: examples/ex33.hpp [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+/**
+ * RationalApproximation_AAA : compute the rational approximation (RA) of
+ * data @a val at the set of points @a pt.
+ *
+ *  INPUT:  @a val       = vector of data values
```
Comment (jamiebramwell) [inline examples/ex33.hpp:47]: Standard doxygen practice for parameters (see [this](https://github.com/mfem/mfem/blob/fe7f097e699d83aa4d545d53a6bba64b80d91146/linalg/ode.hpp#L73) for an example) will improve the resulting doxygen.

**Review by jamiebramwell — APPROVED** [review]
I apologize for my extreme tardiness! Looks great. My only comment is to use standard doxygen formatting for parameters.

File: examples/ex33.hpp (hunk 10 above)
Comment (tzanio) [reply examples/ex33.hpp:47]: @brendankeith , can you address ☝️

Comment (brendankeith) [reply examples/ex33.hpp]: Thanks for pointing that out, @jamiebramwell !

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Really nice example @brendankeith and @psocratis ! 🎉 

@jamiebramwell and @bslazarov -- thank you for your help reviewing it! 🙏

**Pull request merged**